crates/wasi-common/src/entry.rs

```diff
@@ -42,47 +44,97 @@ impl fmt::Debug for Descriptor {
-            Self::VirtualFile(_) => Ok(self),
-            _ => Err(Errno::Badf),
+impl<'descriptor> Deref for OsHandleRef<'descriptor> {
+    type Target = fs::File;
+
+    fn deref(&self) -> &Self::Target {
+        &self.handle
+    }
+}
+
+/// Represents rights of an `Entry` entity, either already held or
+/// required.
+#[derive(Debug)]
+pub(crate) struct EntryRights {
+    pub(crate) base: Rights,
+    pub(crate) inheriting: Rights,
+}
+
+impl EntryRights {
+    pub(crate) fn new(base: Rights, inheriting: Rights) -> Self {
+        Self { base, inheriting }
+    }
+
+    /// Create new `EntryRights` instance from `base` rights only, keeping
+    /// `inheriting` set to none.
+    pub(crate) fn from_base(base: Rights) -> Self {
+        Self {
+            base,
+            inheriting: Rights::empty(),
         }
     }
 
-    /// Return an `OsHandle`, which may be a stream or socket file descriptor.
-    pub(crate) fn as_os_handle<'descriptor>(&'descriptor self) -> OsHandleRef<'descriptor> {
-        descriptor_as_oshandle(self)
+    /// Create new `EntryRights` instance with both `base` and `inheriting`
+    /// rights set to none.
+    pub(crate) fn empty() -> Self {
+        Self {
+            base: Rights::empty(),
+            inheriting: Rights::empty(),
+        }
+    }
+
+    /// Check if `other` is a subset of those rights.
+    pub(crate) fn contains(&self, other: &Self) -> bool {
+        self.base.contains(&other.base) && self.inheriting.contains(&other.inheriting)
+    }
+}
+
+impl fmt::Display for EntryRights {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(
+            f,
+            "EntryRights {{\n\tbase: {},\n\tinheriting: {}\n}}",
+            self.base, self.inheriting
+        )
     }
 }
 
 /// An abstraction struct serving as a wrapper for a host `Descriptor` object which requires
-/// certain base rights `rights_base` and inheriting rights `rights_inheriting` in order to be
-/// accessed correctly.
+/// certain rights `rights` in order to be accessed correctly.
 ///
 /// Here, the `descriptor` field stores the host `Descriptor` object (such as a file descriptor, or
-/// stdin handle), and accessing it can only be done via the provided `Entry::as_descriptor` and
-/// `Entry::as_descriptor_mut` methods which require a set of base and inheriting rights to be
-/// specified, verifying whether the stored `Descriptor` object is valid for the rights specified.
+/// stdin handle), and accessing it can only be done via the provided `Entry::as_descriptor` method
+/// which require a set of base and inheriting rights to be specified, verifying whether the stored
+/// `Descriptor` object is valid for the rights specified.
 #[derive(Debug)]
 pub(crate) struct Entry {
     pub(crate) file_type: Filetype,
-    descriptor: Descriptor,
-    pub(crate) rights_base: Rights,
-    pub(crate) rights_inheriting: Rights,
+    descriptor: Rc<RefCell<Descriptor>>,
```

Comment:
So to prefix this, I'm not very familiar with what this `Descriptor` is and what it's used for. That being said I think it'd be great, if possible, to continue to push the `RefCell` deeper here. The purpose of removing the ability to get `&mut Entry` was all about making mutable borrows as small as possible to eliminate possibilities of panics, and I think the same might apply here as well. It may even be possible to have `Descriptor` only have `Cell` fields, but ideally any `RefCell` fields would have very small borrows and no methods would have to return a `Ref` or `RefMut`

Reply:
I think you're right. Although making it work will require some bigger reorg to `OsHandle` especially on BSD platform. I have an idea for this and will investigate but given that this probably will be a bigger change, I'd prefer to have it as a subsequent PR. In short, if we make `OsHandle` a bit closer to the metal by replacing `File` with `RawFd`, we can make `OsHandle` `Copy` which would then make it possible to have `Descriptor` inside a `Cell`. There's a little bit more work involved with BSD since currently `OsHandle`, in addition to `File`, also stores an optional `yanix::Dir` (which is non-`Copy` since it's got a custom `Drop` impl) and it would probably make sense to split the handle into either representing an actual file, or a directory. Then, we wouldn't need `Option` nor interior mutability, since we could easily make `yanix::Dir` `Copy` calling it something like `yanix::UnsafeDir` and specifically mentioning that it is the user's responsibility to close the underlying dir stream with `libc::closedir` fn. Does that make sense?

Reply:
Oh, and the only reason we ever need to mutate the actual `Descriptor` in `Entry` is because on Windows, setting fdstat flags requires reopening the file.

Reply:
Ah ok punting this to later seems fine, in any case looks like a great refactoring!